[16.0][ADD] shopfloor_packing

sbejaoui opened the pull request.

sbejaoui committed:
[16.0][ADD] - shopfloor_packing

jbaudoux commented on shopfloor_packing/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Shopfloor Packing",
```
As this is an improvement of the cluster picking scenario, can this be renamed something like "shopfloor_cluster_picking_repack" ?

rousseldenis replied:
Maybe in a further version as too much implications in change.

lmignon commented on shopfloor_packing/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,2 @@
+This module adds a new configuration option to the "cluster picking"
```
@sbejaoui Can you explain more precisely why and when you could use this addon

sbejaoui committed:
[IMP] description

sbejaoui committed:
[IMP] add failing test for packing partial qty

If a partially available product becomes completely available and the reserved
quantity increases after the "put in pack" action, a new move line will be
created with the done quantity as a copy of the line that will be packed.
The original line will keep the result package, which is the package scanned
by the operator, and the remaining quantity to do.

Since the package is an internal package, the line with the remaining quantity
to do will be proposed in the packing process and considered as picked. If the
operator don't pay attention, and confirm the packaging, both lines will be set
to done.

sbejaoui committed:
[FIX] packing partial qty

rousseldenis committed:
[IMP] shopflor/mobile/_packing: Allows to select package types for packing

rousseldenis committed:
[IMP] shopfloor_packing: Move data retrieval to data action + enable tests

rousseldenis committed:
[IMP] shopfloor_packing: Use better message

jbaudoux commented on shopfloor_packing/models/stock_picking_batch.py:
```diff
@@ -0,0 +1,21 @@
+
+    is_shopfloor_packing_todo = fields.Boolean(
+        "Operations need to be packed",
+        help="If set, some operations need to be packed by the shopdloor operator",
```
```suggestion
        help="If set, some operations need to be packed by the shopfloor operator",
```

jbaudoux commented on shopfloor_packing/models/shopfloor_menu.py:
```diff
@@ -0,0 +1,15 @@
+        string="Pack pickings",
+        default=False,
+        help="If you tick this box, all the picked item will be put in pack"
+        " before the transfer.",
```
```suggestion
    pack_pickings = fields.Boolean(
        string="Pack pickings",
        default=False,
        help="If you tick this box, all the picked items will be put in a new pack"
        " at the end of the picking process.",
```

rousseldenis committed:
[IMP] shopfloor_packing: Use appropriate name for packing action + improve tests

rousseldenis committed:
[IMP] shopfloor_packing: Several cosmetics + update list_delivery_packaging endpoint + add tests

rousseldenis committed:
[IMP] shopfloor_mobile_packing: Update list_delivery_packaging endpoint

rousseldenis committed:
[IMP] shopfloor_packing: Update docstring for package types

rousseldenis committed:
[IMP] shopfloor_mobile_packing: Improve docstrings + add contributors

rousseldenis committed:
[IMP] shopfloor_packing: Improve helps + add contributors

rousseldenis commented:
@jbaudoux

jbaudoux commented on shopfloor_mobile_packing/static/src/js/i18n/en.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "cluster_picking.pack_picking_put_in_pack.title": "Packing: Enter the number of boxes used.",
+  "cluster_picking.pack_picking_put_in_pack.scan_placeholder": "Number of boxes used",
```
As now we choose a package type, this needs to be adapted.
```suggestion
  "cluster_picking.pack_picking_put_in_pack.title": "Packing",
  "cluster_picking.pack_picking_put_in_pack.scan_placeholder": "Package type",
```

jbaudoux commented on shopfloor_mobile_packing/static/src/js/i18n/fr.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "cluster_picking.pack_picking_put_in_pack.title": "Emballage: Nombre de colis.",
+  "cluster_picking.pack_picking_put_in_pack.scan_placeholder": "Nombre de colis créés?",
```
same here

jbaudoux commented on shopfloor_packing/i18n/fr_BE.po:
```diff
@@ -0,0 +1,95 @@
+# Translation of Odoo Server.
```
`fr.po` instead of `fr_BE.po`

jbaudoux commented on shopfloor_packing/i18n/fr_BE.po:
```diff
@@ -0,0 +1,95 @@
+# Translation of Odoo Server.
+# This file contains the translation of the following modules:
+#	* alc_shopfloor_packing
```
replace all or regenerate file
```suggestion
#	* shopfloor_packing
```

jbaudoux commented on shopfloor_packing/i18n/fr_BE.po:
```diff
@@ -0,0 +1,95 @@
+#: model:ir.model.fields,help:alc_shopfloor_packing.field_stock_picking_is_shopfloor_packing_todo
+#: model:ir.model.fields,help:alc_shopfloor_packing.field_stock_picking_wave_is_shopfloor_packing_todo
+msgid "If set, some operations need to be packed by the shopdloor operator"
+msgstr "Si coché, le contenu des bacs doit encore être mis en boîte par l'opérateur Shofloorr"
```
```suggestion
msgid "If set, some operations need to be packed by the operator"
msgstr "Si coché, le contenu des bacs doit encore être mis en boîte par l'opérateur"
```

jbaudoux commented on shopfloor_packing/i18n/fr_BE.po:
```diff
@@ -0,0 +1,95 @@
+#. module: alc_shopfloor_packing
+#: code:addons/alc_shopfloor_packing/actions/message.py:21
+#, python-format
+msgid "The number of packages must be greater than 0."
+msgstr "Le nombre de boîtes doit est > 0."
```
likely obsolete

rousseldenis replied:
Nope as you can choose behavior

jbaudoux commented on shopfloor_packing/models/shopfloor_menu.py:
```diff
@@ -0,0 +1,23 @@
+        " at the end of the picking process.",
+    )
+
+    default_pack_pickings_action = fields.Selection(
+        [
+            ("nbr_packages", "Enter the number of packages"),
+            ("package_type", "Scan the package type"),
+        ],
+        default="nbr_packages",
+    )
```
This is obsolete, isn't it?

rousseldenis replied:
You can choose behavior

jbaudoux reviewed (commented):
quick review

jbaudoux commented on shopfloor_mobile_packing/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    "summary": """
+    Shopfloor mobile extension for packing operation into cluster picking
+    """,
+    "author": "ACSONE SA/NV,Odoo Community Association (OCA)",
```
```suggestion
    "author": "ACSONE SA/NV,BCIM,Odoo Community Association (OCA)",
```

jbaudoux commented on shopfloor_packing/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    "name": "Shopfloor Packing",
+    "version": "16.0.1.0.0",
+    "summary": """ Manage Packing into cluster picking""",
+    "author": "ACSONE SA/NV,Odoo Community Association (OCA)",
```
```suggestion
    "author": "ACSONE SA/NV,BCIM,Odoo Community Association (OCA)",
```

rousseldenis committed:
[IMP] shopfloor_packing: Move BE translation to generic fr one

rousseldenis committed:
[IMP] shopfloor_*_packing: Improve packing flow

lmignon committed:
[IMP] shopfloor_packing: Avoid concurrent update

Simlify the content of the message returned when packing a picking or selecting a package. The most important change it to not include into the product's information the qty_available field. The computation of this useless info in such context could lead to concurrent update since it requires to select all the quants into the stock while a lot of operations on the quants are done in parallel.

lmignon committed:
[FIX] shopfloor_packing: Ensure number of parcels is preserved

if a number of parcels is specified and a package type with a different default value is used, we must ensure that at the end of the process the encoded number of parcels is the one specified into the UI

sbejaoui committed:
[IMP] shopfloor_packing: compute result package weight through move lines

there are two methods to compute the result package estimated weight: using move lines or using quants

using move lines is more efficient and avoids concurrent access on stock quants.
to enable this approach, the picking_id must be set in the context

sbejaoui committed:
[IMP] shopfloor_packing: skip package quantity computation in packing scenario

simahawk commented on shopfloor_packing/models/stock_picking.py:
```diff
@@ -0,0 +1,78 @@
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    is_shopfloor_packing_todo = fields.Boolean(
+        "Operations need to be packed",
+        help="If set, some operations need to be packed by the shopdloor operator",
+        compute="_compute_is_shopfloor_packing_todo",
+    )
+
+    packing_scanned_packs_str = fields.Char(
+        help="Technical field to know which pack has been scanned into the put in "
+        "pack process"
+    )
```
```suggestion
    shopfloor_packing_todo = fields.Boolean(
        "Shopfloor: pack after picking",
        help="If set, these transfer will be packed at the end of the picking done via Shopfloor app.",
        compute="_compute_shopfloor_packing_todo",
    )

    shopfloor_packing_scanned_packs = fields.Json(
        help="Technical field to know which pack has been scanned into the put in "
        "pack process"
    )
```

simahawk commented on shopfloor_packing/models/stock_picking.py:
```diff
@@ -0,0 +1,78 @@
+                    break
+
+    @property
+    def _packing_scanned_packs(self):
```
why not a serialized field then?

simahawk commented on shopfloor_packing/models/stock_picking.py:
```diff
@@ -0,0 +1,78 @@
+
+    def _put_in_pack(self, move_line_ids, create_package_level=True):
+        """
+        Reset the result package after partial put in pack
```
this is a comment

simahawk commented on shopfloor_packing/models/stock_picking_batch.py:
```diff
@@ -0,0 +1,21 @@
+class StockPickingBatch(models.Model):
+    _inherit = "stock.picking.batch"
+
+    is_shopfloor_packing_todo = fields.Boolean(
```
same comments as for the other field

simahawk commented on shopfloor_packing/services/packing.py:
```diff
@@ -0,0 +1,78 @@
+from odoo.addons.stock.models.stock_picking import Picking
+
+
+class PackingAction(Component):
```
this goes to actions folder

simahawk commented on shopfloor_packing/models/shopfloor_menu.py:
```diff
@@ -0,0 +1,23 @@
+        " at the end of the picking process.",
+    )
+
+    default_pack_pickings_action = fields.Selection(
```
```suggestion
    pack_pickings_default_action = fields.Selection(
```

simahawk commented on shopfloor_packing/models/stock_picking.py:
```diff
@@ -0,0 +1,78 @@
+        scanned_packs = list(packing_scanned_packs) if packing_scanned_packs else []
+        self.packing_scanned_packs_str = json.dumps(scanned_packs)
+
+    def _set_packing_pack_scanned(self, pack_id):
```
IMO this API should go to the packing action

simahawk commented on shopfloor_mobile_packing/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2021 ACSONE SA/NV
```
this module should be renamed to `shopfloor_packing_mobile`

simahawk commented on shopfloor_mobile_packing/static/src/js/cluster-picking.js:
```diff
@@ -0,0 +1,352 @@
+// Keep the pointer to the orginal method
+let data_result_method = ClusterPickingBase.component.data;
+
+ClusterPickingBase.component.computed.searchbar_input_type = function () {
```
AFAIR there's a better way to override these functions....

simahawk commented on shopfloor_mobile_packing/static/src/js/cluster-picking.js:
```diff
@@ -0,0 +1,352 @@
+    // the object context
+    let result = data_result_method.bind(this)();
+    // add our new state
+    result.states.pack_picking_put_in_pack = {
```
IMO this part should be done like what we've done for checkout: split states definition out and override `_get_states` method.

simahawk commented on shopfloor_mobile_packing/static/src/js/cluster-picking.js:
```diff
@@ -0,0 +1,352 @@
+        },
+        on_select: (selected) => {
+            return;
+            // TODO:
```
?

simahawk commented on shopfloor_mobile_packing/static/src/js/cluster-picking.js:
```diff
@@ -0,0 +1,352 @@
+            // );
+        },
+        on_back: () => {
+            $instance.state_to("select_line");
```
this is broken

simahawk commented on shopfloor_packing/actions/data.py:
```diff
@@ -0,0 +1,61 @@
+            "id": record.id,
+            "name": record.name,
+            "partner": {"id": record.partner_id.id, "name": record.partner_id.name},
+            "scanned_packs": list(record._packing_scanned_packs),
```
```suggestion
            "scanned_packs": record.shopfloor_scanned_packs,
```
If that field gets renamed and becomes a seraliazed field

simahawk commented:
FYI we need to work on this for v18. I'll do some cleanup and try to fix the issues I've pointed out here.
At the end I'll squash commits by module and keep authorship of course.
Then we might backport the final version from v18 if it gets simpler.

rousseldenis committed:
[FIX] shopfloor_mobile_packing: Use correct data member for scan_package_action()

jbaudoux commented:
Once https://github.com/OCA/wms/pull/1218 is merged,
you will need to cherry-pick https://github.com/OCA/stock-logistics-shopfloor/pull/88/changes/aa315f5f12402388b888563a9e60a9f040e42e68 from https://github.com/OCA/stock-logistics-shopfloor/pull/88 to fix the test

lmignon commented:
ping @sbejaoui
